### PR TITLE
[WIP] use horde sync to keep messages and message lists up to date

### LINKED ISSUE #733: Detect flag changes and expunged messages
ATM we only check for new messages (uid validity, next uid). Additionally, it would be great to detect and sync any other changed flags (unseen, starred, replied, deleted,...) or messages deleted in other IMAP clients.

:books:  https://tools.ietf.org/html/rfc4549#section-4.3 seems to be a good strategy to do that

Related: #663

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes #733

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
Nice start - thx. Just perfect timing - I need this fir the unified inbox as well.

File: js/mail.js [hunk 1]
```diff
@@ -119,6 +119,38 @@ var Mail = {
...
+		var syncTokens = {};
+		// Update every 15 seconds
+		var updateInterval = 15;
+		var timer = null;
+
+		function detectChanges() {
+			var accountId = Mail.State.currentAccountId;
+			var folderId = Mail.State.currentFolderId;
+			// Only start if there ary any accounts/folders/tokens
+			if (accountId && folderId && syncTokens > 0) {
```
Comment (wurstchristoph) [inline js/mail.js]: TODO: syncTokens.length

**Comment by wurstchristoph** [comment]
@DeepDiver1975 I'll need your help on this. I always get an exception [here](https://github.com/horde/horde/blob/master/framework/Imap_Client/lib/Horde/Imap/Client/Data/Sync.php#L157).

Are there any preconditions to make sync work? There's very little documentation on that :/

**Comment by DeepDiver1975** [comment]
I will debug into this the next days. Promised.

**Comment by wurstchristoph** [comment]
Some further trial and error shows that sync __does__ work when used right after requesting the token. I could successfully detect any changes [here](https://github.com/owncloud/mail/pull/763/files#diff-ee701d435b416eddecc1798126969f4dR125).

Moreover I could find some code using the sync functionality: https://github.com/horde/horde/blob/3aff0b0dfcfb190e307b3dc01edfa87d4baaacc9/imp/lib/Ajax/Application/ListMessages.php#L384

I'm not totally sure, but it could be that we need server-side caching (I assume horde needs to know the previous mailbox state to detect any changes). I enabled caching in my dev env using `` 'app.mail.server-side-cache.enabled' => true``, but this didn't help neither :(

File: appinfo/routes.php [hunk 2]
```diff
@@ -32,6 +32,10 @@
 				'url' => '/accounts/{accountId}/folders/{folderId}/messages/{messageId}/toggleStar',
 				'verb' => 'POST'],
 			[
+				'name' => 'messages#detectChanges',
```
Comment (DeepDiver1975) [inline appinfo/routes.php]: why is there another route? I don't get this - sorry

File: lib/controller/messagescontroller.php [hunk 3]
```diff
@@ -332,4 +362,14 @@ private function buildHtmlBodyUrl($accountId, $folderId, $id) {
...
 	}
 
+	/**
+	 * 
+	 * @param \OCA\Mail\Mailbox $mailBox
+	 */
+	private function getSyncToken(Mailbox $mailBox) {
+		$m = new Account($this->getAccount());
+		$client = $m->getImapConnection();
+		return $client->getSyncToken($mailBox->getHordeMailBox());
```
Comment (DeepDiver1975) [inline lib/controller/messagescontroller.php]: Oh - the sync token is mailbox based? Hmmmm ....

File: appinfo/routes.php (hunk 2 above)
Comment (wurstchristoph) [inline appinfo/routes.php]: > why is there another route? I don't get this - sorry

Will be merged with the existing ``folders#detectChanges`` route, I added this just to try things out ;-)

**Pull request closed without merging**